pf/internal/convert/encoding.go

```diff
@@ -268,15 +269,30 @@ func (e *encoding) deriveEncoder(typeSpec *pschema.TypeSpec, t tftypes.Type) (En
 	case "string":
 		return newStringEncoder(), nil
 	case "array":
-		lt, ok := t.(tftypes.List)
-		if !ok {
-			return nil, fmt.Errorf("expected a List, got %s", t.String())
-		}
-		elementEncoder, err := e.deriveEncoder(typeSpec.Items, lt.ElementType)
-		if err != nil {
-			return nil, err
+		switch t := t.(type) {
+		case tftypes.List:
```

Comment:
I shied away from objects, since that would require naming, and `(P1_T1, P1_T2) != (P2_T1, P2_T2)` even if `P1_T1 = P2_T1 = string` and `P1_T2 = P2_T2 = int`, which will be confusing (either at runtime or compile time, depending on the language). I'll try an object iteration on this, since I agree it preserves type information while lists do not, but I'm concerned python users will complain that `pkg.mod.StringIntTuple("string", 0)` is absurdly verbose compared to `"string", 0`.

Reply:
Well Python has native tuples, so perhaps a feature request we could make upstream to the Pulumi schema to support tuple types natively and map them to Python tuples, or perhaps more precariously to magic-recognize encoded tuples in the Python encoding so that Python tuples are supported as tuples.

Reply:
The request would need some motivating examples though.. So far I've not seen this feature widely used, but we need more providers. From what we know today it sounds like preserving type info in the projection is worth it and worth some awkwardness, with the awkwardness not affecting too many users because it seems like a rare feature.